Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp

```diff
@@ -316,38 +293,127 @@ static constexpr SDLMotionAxisList SDL_AXES_GYRO = {{
+      is_button_mapped.assign(bind.value.button, true);
+      break;
+    case SDL_CONTROLLER_BINDTYPE_AXIS:
+      is_axis_mapped.assign(bind.value.axis, true);
+      break;
+    case SDL_CONTROLLER_BINDTYPE_HAT:
+      is_hat_mapped.assign(bind.value.hat.hat, true);
+      break;
+    }
+  };
+
+  if (gamecontroller != nullptr)
   {
-    // This device is invalid, don't use it
-    // Some crazy devices(HP webcam 2100) end up as HID devices
-    // SDL tries parsing these as joysticks
-    return;
+    // Inputs
+
+    // Buttons
+    for (u8 i = 0; i != size(s_sdl_button_names); ++i)
+    {
+      SDL_GameControllerButton button = static_cast<SDL_GameControllerButton>(i);
+      if (SDL_GameControllerHasButton(m_gamecontroller, button))
+      {
+        AddInput(new Button(gamecontroller, button));
+        RegisterMapping(SDL_GameControllerGetBindForButton(gamecontroller, button));
+      }
+    }
+
+    // Axes
+    for (u8 i = 0; i != size(s_sdl_axis_names); ++i)
+    {
+      SDL_GameControllerAxis axis = static_cast<SDL_GameControllerAxis>(i);
+      if (SDL_GameControllerHasAxis(m_gamecontroller, axis))
+      {
+        // First 4 axes are for the analog sticks, the rest are for the triggers
+        if (i < 4)
+        {
+          // Each axis gets a negative and a positive input instance associated with it
+          AddAnalogInputs(new Axis(m_gamecontroller, -32768, axis),
+                          new Axis(m_gamecontroller, 32767, axis));
+        }
+        else
+        {
+          AddInput(new Axis(m_gamecontroller, 32767, axis));
+        }
+        RegisterMapping(SDL_GameControllerGetBindForAxis(gamecontroller, axis));
+      }
+    }
+    // Rumble
+    if (SDL_GameControllerHasRumble(m_gamecontroller))
+    {
+      AddOutput(new Motor(m_gamecontroller));
+    }
+
+    // Motion
+    auto AddSensor = [this](SDL_SensorType type, std::string_view name,
+                            const SDLMotionAxisList& axes) {
+      if (SDL_GameControllerSetSensorEnabled(m_gamecontroller, type, SDL_TRUE) == 0)
+      {
+        for (const SDLMotionAxis& axis : axes)
+        {
+          AddInput(new MotionInput(fmt::format("{} {}", name, axis.name), m_gamecontroller, type,
+                                   axis.index, axis.scale));
+        }
+      }
+    };
+
+    AddSensor(SDL_SENSOR_ACCEL, "Accel", SDL_AXES_ACCELEROMETER);
+    AddSensor(SDL_SENSOR_GYRO, "Gyro", SDL_AXES_GYRO);
+    AddSensor(SDL_SENSOR_ACCEL_L, "Accel L", SDL_AXES_ACCELEROMETER);
+    AddSensor(SDL_SENSOR_GYRO_L, "Gyro L", SDL_AXES_GYRO);
+    AddSensor(SDL_SENSOR_ACCEL_R, "Accel R", SDL_AXES_ACCELEROMETER);
+    AddSensor(SDL_SENSOR_GYRO_R, "Gyro R", SDL_AXES_GYRO);
   }
 
-  // get buttons
-  for (u8 i = 0; i != SDL_JoystickNumButtons(m_joystick); ++i)
-    AddInput(new Button(i, m_joystick));
+  // Legacy inputs
+
+  // Buttons
+  for (u8 i = 0; i != n_legacy_buttons; ++i)
+    AddInput(new LegacyButton(m_joystick, i, !is_button_mapped[i]));
```

Comment:
Instead of adding "not detectable" legacy inputs. I think it would be cleaner for the newly introduced GameController `Input`s to override `IsMatchingName` to match the old "Button 1" naming. This would still allow loading old configs but without flooding the input list with redundant entries. This is how input backend naming changes have been handled in the past.

Reply:
I tried that, but doing that is a mess.
Just because something is a Button in the GameController API doesn't mean it's a Button in the Joystick API too.
Axes have different ranges (-32768 to 32767 normally, but are from 0 to 32767 *only* in the GameController API).

It's not worth it.

All the legacy inputs are added at the end, so that they are at the bottom of the list.